### PR TITLE
[_transactions2] Coordination, Part 5: Async Initializable Coordination Store

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Make the coordination store async initializable
- Transactions2 / Transactions Service eventually needs to follow the rules of async initialization.

**Implementation Description (bullets)**:
- Implement standard async initialization primitives for `KeyValueServiceCoordinationStore`

**Testing (What was existing testing like?  What have you done to improve it?)**:
There is one new test that handles the case where the KVS isn't ready initially, but becomes ready

**Concerns (what feedback would you like?)**: Nothing in particular

**Where should we start reviewing?**: `KeyValueServiceCoordinationStore.java`

**Priority (whenever / two weeks / yesterday)**: This week, hopefully? If not early next

### CODE REVIEW

File: atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationStore.java [hunk 1]
```diff
@@ -20,11 +20,23 @@
...
  * A {@link CoordinationStore} stores data that a {@link CoordinationService} may use.
  */
+@AutoDelegate
 public interface CoordinationStore<T> {
+    /**
+     * Coordination stores may require asynchronous initialization, if dependencies aren't initially available.
+     *
+     * @return true iff the coordination store is ready to service requests
+     */
+
```
Comment (sandorw) [inline atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationStore.java]: nit: empty line

File: atlasdb-coordination-impl/src/test/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStoreTest.java [hunk 2]
```diff
@@ -159,10 +169,31 @@ public void multipleStoresCanCoexist() {
...
+                .createTable(any(), any());
+
+        CoordinationStore<String> asyncStore = KeyValueServiceCoordinationStore.create(
+                ObjectMappers.newServerObjectMapper(),
+                kvs,
+                COORDINATION_ROW,
+                timestampSequence::incrementAndGet,
+                String.class,
+                true);
+        Awaitility.waitAtMost(Duration.ONE_MINUTE)
```
Comment (sandorw) [inline atlasdb-coordination-impl/src/test/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStoreTest.java]: Is this how we generally test this elsewhere? The alternative would be to just call `tryInitialize` directly, assuming the general async initialization stuff is tested independently.

What's the default wait period for async init? It looks like the default behavior is to wait 10 seconds.

File: atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStore.java [hunk 3]
```diff
@@ -319,4 +332,21 @@ private static TableMetadata getCoordinationTableMetadata() {
                 false,
                 TableMetadataPersistence.LogSafety.SAFE);
     }
+
+    private class InitializingWrapper extends AsyncInitializer implements AutoDelegate_CoordinationStore<T> {
```
Comment (gmaretic) [inline atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStore.java]: The convention we used is to put this as the first item in the class

File: atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationStore.java (hunk 1 above)
Comment (gmaretic) [inline atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationStore.java]: nit: newline

Comment (jeremyk-91) [reply atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationStore.java]: Good catch, removed.

File: atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStore.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStore.java]: Alright, have brought it to the front.

File: atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationStore.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationStore.java]: Yep, fixed - interesting that this rendered separately from Sandor's comment

File: atlasdb-coordination-impl/src/test/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStoreTest.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-coordination-impl/src/test/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStoreTest.java]: Right, the default wait period is 10s. I agree that it's probably better to hit `tryInitialize` for this test, to keep test duration short and avoid waiting 10s - also avoid failures if something else in async init like the standard wait period changes.

Not aware of how this is tested elsewhere apart from integration tests - e.g. `LockStoreImpl`, another use of this, is untested.

Comment (jeremyk-91) [reply atlasdb-coordination-impl/src/test/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStoreTest.java]: Post offline discussion - Not sure if this test brings much value, as the framework is tested elsewhere, and this is a straightforward usage (no cleanup tasks), so removing this.

Also taking out Awaitility as we don't use it elsewhere

**Pull request merged**